File: src-tauri/src/services/structure/rules.rs
use super::exceptions::{
    InvalidDirectoryIdException, InvalidDirectoryNameException, InvalidParentException,
    InvalidPositionException, InvalidResourceNameException, InvalidStructureNodeIdException,
    InvalidStructureResourceIdException,
};

pub struct StructureRules;

impl StructureRules {
    pub fn validate_node_id(id: &str) -> Result<String, InvalidStructureNodeIdException> {
        let normalized = id.trim().to_string();
        if normalized.is_empty() {
            return Err(InvalidStructureNodeIdException {
                message: "Node id must not be empty.".into(),
            });
        }
        Ok(normalized)
    }

    pub fn validate_resource_id(id: &str) -> Result<String, InvalidStructureResourceIdException> {
        let normalized = id.trim().to_string();
        if normalized.is_empty() {
            return Err(InvalidStructureResourceIdException {
                message: "Resource id must not be empty.".into(),
            });
        }
        Ok(normalized)
    }

    pub fn validate_directory_id(id: &str) -> Result<String, InvalidDirectoryIdException> {
        let normalized = id.trim().to_string();
        if normalized.is_empty() {
            return Err(InvalidDirectoryIdException {
                message: "Directory id must not be empty.".into(),
            });
        }
        Ok(normalized)
    }

    pub fn validate_directory_name(name: &str) -> Result<String, InvalidDirectoryNameException> {
        let normalized = name.trim().to_string();
        if normalized.is_empty() {
            return Err(InvalidDirectoryNameException {
                message: "Directory name must not be empty.".into(),
            });
        }
        Ok(normalized)
    }

    pub fn validate_position(position: i64) -> Result<i64, InvalidPositionException> {
        if position < 0 {
            return Err(InvalidPositionException {
                message: "Position must not be negative.".into(),
            });
        }
        Ok(position)
    }

    pub fn validate_course_id(id: &str) -> Result<String, InvalidParentException> {
        let normalized = id.trim().to_string();
        if normalized.is_empty() {
            return Err(InvalidParentException {
                message: "Course id must not be empty.".into(),
            });
        }
        Ok(normalized)
    }

    pub fn validate_resource_name(name: &str) -> Result<String, InvalidResourceNameException> {
        let normalized = name.trim().to_string();
        if normalized.is_empty() {
            return Err(InvalidResourceNameException {
                message: "Resource name must not be empty.".into(),
            });
        }
        Ok(normalized)
    }
}

File: src/entities/structure/core/rules.ts
import {
  InvalidDirectoryIdError,
  InvalidDirectoryNameError,
  InvalidParentError,
  InvalidPositionError,
  InvalidStructureNodeIdError,
  InvalidStructureResourceIdError,
  InvalidStructureResourceNameError,
} from './exceptions'

type EntityError = new (message: string) => Error
function id(value: string, error: EntityError, label: string): string {
  const normalized = value.trim()
  if (!normalized) throw new error(`${label} не может быть пустым`)

  return normalized
}
export const validateNodeId = (value: string) =>
  id(value, InvalidStructureNodeIdError, 'Идентификатор узла')
export const validateResourceId = (value: string) =>
  id(value, InvalidStructureResourceIdError, 'Идентификатор ресурса')
export const validateDirectoryId = (value: string) =>
  id(value, InvalidDirectoryIdError, 'Идентификатор директории')
export const validateCourseId = (value: string) =>
  id(value, InvalidParentError, 'Идентификатор курса')
export const validateDirectoryName = (value: string) =>
  id(value, InvalidDirectoryNameError, 'Название директории')
export const validateStructureResourceName = (value: string) =>
  id(value, InvalidStructureResourceNameError, 'Название ресурса')
export function validatePosition(value: number): number {
  if (value < 0) throw new InvalidPositionError('Позиция не может быть отрицательной')

  return value
}
